src/Psl/Collection/MutableSet.php

```diff
@@ -0,0 +1,716 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Collection;
+
+use Closure;
+use Psl\Dict;
+use Psl\Iter;
+use Psl\Vec;
+
+use function array_key_exists;
+use function array_key_first;
+use function array_key_last;
+use function array_keys;
+use function count;
+
+/**
+ * @template T of array-key
+ *
+ * @implements MutableSetInterface<T>
+ */
+final class MutableSet implements MutableSetInterface
+{
+    /**
+     * @var array<T, T>
+     */
+    private array $elements = [];
+
+    /**
+     * MutableSet constructor.
+     *
+     * @param array<T, mixed> $elements
+     *
+     * @psalm-mutation-free
+     */
+    public function __construct(array $elements)
+    {
+        $set = [];
+        foreach ($elements as $key => $_) {
+            $set[$key] = $key;
+        }
```

Comment:
Per our conversation offline, I feel it would be more intuitive to use the values of `$elements`, rather than the keys, to construct a MutableSet – something like this:
```suggestion
     * @param array<array-key, T> $elements
     *
     * @psalm-mutation-free
     */
    public function __construct(array $elements)
    {
        $set = [];
        foreach ($elements as $_ => $value) {
            $set[$value] = $value;
        }
```

This would also have implications for other method implementations below, like `map()`.